### ISSUE TITLE
Modulo bias in `randint`

### ISSUE THREAD

**Issue by pierce314159** [Bears-R-Us/arkouda#1220]
`ak.randint(aMin, aMax)` is intended to generate uniformly distributed integers in the `range(aMin, aMax)`. Since we are using a modulo to convert randomly generated `uint64`s to a number in our desired range, we are systematically favoring smaller values when `aMax-aMin` does not evenly divide `2**64`

This was pointed out by brad in PR #1214
> If I'm generating random 8-bit sequences and then using `%` to try and constrain them to be between 0..250, statistically speaking, I'm going to get twice as many entries in the 0..5 range than the 6..250 range due to wraparound, aren't I?  I think the more typical practice is to drop out-of-bound or "leftover" values on the floor and "re-roll"

_Originally posted by @bradcray in https://github.com/Bears-R-Us/arkouda/pull/1214#discussion_r828219173_

**Comment by pierce314159** [comment]
I believe our proposed path forward on this is to have a member of the chapel team write code allowing min/max arguments to be passed into `fillRandom`. Then we can drop that in as a workaround function until that functionality is in our lowest supported chapel version

**Comment by bradcray** [comment]
> This appears to be handled by PR https://github.com/chapel-lang/chapel/pull/19540

That's correct.  This hasn't been merged yet due to distractions from releases and outages, but I think should be ready to go.

**Comment by bradcray** [comment]
As an update, https://github.com/chapel-lang/chapel/pull/19540 is merged to Chapel's `main` branch now.

**Comment by Ethan-DeBandi99** [comment]
Hoping to get a bit of guidance from the chapel team on this one. I have been looking into moving the functionality from the PR mentioned in previous comments into Arkouda as a temporary solution to this issue (until our lowest supported version is updated), but am hitting some snags. I am hoping this is just me missing something. @pierce314159 recommended I try adding the functions to the module using something like 

```chapel
use Random;

 proc Random.fillRandomWorkaround(arr: [], min: arr.eltType, max:arr.eltType) {
        if(!arr.isRectangular()) then
          compilerError("fillRandom does not support non-rectangular arrays");

        forall (x, r) in zip(arr, iterate(arr.domain, arr.eltType,
                                          min, max)) do
          x = r;
      }
```

It is very possible I missed one of the updates or did not do this properly, but it did not seem to work. I also tried the brute force method of creating a file `AKRandom.chpl` and copying `Random.chpl` into it. I renamed the modules and a few other things. It seems as though it is not finding the updates, almost as if it is using `Random` even though I updated to `use AKRandom`. Hopefully, this is just something simple I have yet to encounter with Chapel, but would greatly appreciate any pointers or recommendations anyone may have.

@bradcray @mppf @ronawho @bmcdonald3

**Comment by bradcray** [comment]
@Ethan-DeBandi99 : I think the issue here is the `Random.` on your `proc Random.fillRandomWorkaround()` declaration.  Specifically, there isn't a way in Chapel to inject code into an existing module's namespace (which is what this syntax suggests to me).  But if you just had `proc fillRandomWorkaround()` I would expect it to work fine, at least thinking about it quickly.  If not, let us know what kinds of problems you're hitting.

**Comment by Ethan-DeBandi99** [comment]
@bradcray that was just a quick Audi codeish example that @pierce314159 had given me as an example. I am off for the day, but I will get the specifics for you Monday morning. I would venture to guess it is almost certainly something I am doing wrong.

**Comment by pierce314159** [comment]
I think I might be to blame here. I was going off an example that @e-kayrakli was helping me with back with the regex stuff (my only time dropping in a workaround from the chapel team). I wasn't too familiar with chapel at the time but interpreted this as adding `subnWorkaround` to the `Regex` namespace, but now I think it's adding this method to the `record regex` from `module Regex`

```chapel
use Regex;

var myRegex = compile(b"a+");

var str = b"oneatwo";

proc regex.subnWorkaround(repl: bytes, text: bytes, global=true): (bytes, int) {
  var _repl = repl;
  if repl == b"\x00" then
    _repl = b"\xff";
  var (res, n) = this.subn(_repl, text, global);

  return (res.replace(_repl, repl), n);
}

var (res, n) = myRegex.subnWorkaround(b"\x00", str);

writeln(res);
writeln(res.size);
writeln(n);
```

This lines up with what we do in `Flatten`, where we add a method to `SegString`
https://github.com/Bears-R-Us/arkouda/blob/b11e22bb1176c553bf86d08ecf375996d5a28a2c/src/Flatten.chpl#L33

Let me know if I'm off base again. Also @bradcray, do you know of a better way to do this other than just copy the functions that changed  in of `Random` and update any references to those to be the `*Workaround` versions?

**Comment by bradcray** [comment]
I think you're right @pierce314159.  New methods can be defined on types outside of that type's declaration or module itself ("tertiary methods"), but we don't support this on modules themselves.

Totally copying the file would work, though it'd impose a slight maintenance burden in that, at some point in the future (after 1.27), you'd want to remove it and just use `Random` itself, which would require remembering to do so.  Otherwise, you'd start to lose any new features or benefits added to `Random` in the intervening time.  I think that if I were doing this, I would: 

* create compatibility directories `compat/lt-127` `compat/ge-127`
* create a new module in `compat/lt-127` named `AKRandom.chpl` or `RandomHelper.chpl` or whatever you like that contains the new routine which is not available in 1.26, using the same name as on `main` today.  This module could also contain `public use Random;` if you only wanted to `use` it, or `[private] use Random;` if you wanted Arkouda to `use` both `Random` and this helper module (I'd probably do the latter to keep it from seeming to subsume Random).
* create an empty module with the same name in `compat/ge-127` (or, if the previous module contained `public use Random;` you'd want this one to as well for consistency).

Then you should have a situation where any version of the compiler will find the routine with the same name; and where once no versions older than 1.27 are supported with Arkouda anymore, you can just remove these modules and the `use` of them.

**Comment by mppf** [comment]
Note that if you rename the new Random to AKRandom you will probably also have to similarly rename submodules contained within it.

**Comment by Ethan-DeBandi99** [comment]
@mppf - I had tried going through and doing that. There is a potential that I missed something. I am working through @bradcray suggestion now. Unfortunately, I am seeing similar issues with it not finding some of the function definitions.

**Comment by bradcray** [comment]
@Ethan-DeBandi99 : Would you say what the not-found function definitions are?  My guess is that they're defined in a sub-module of `Random` and that you'll need to `use` that/those submodules as well (?).

**Comment by Ethan-DeBandi99** [comment]
@bradcray  - They are indeed. In fact, most of the `helpers` we are overriding with are in the submodules within `Random`. So I created the directories you recommended and have the file set up something like

```chapel
module AKRandom {
    public use Random; //This is what the other instances within compat use

    module AK<submodule_name> {
        //Function overrides
    }
}
```
And then the file using `Random` is changed to `use AKRandom`. I apologize for stumbling through this. It seems fairly straightfoward, but I am obviously missing something.

**Comment by bradcray** [comment]
@Ethan-DeBandi99 : I tried to see if I could put my own prototype together since it seemed like it should be easier than it was sounding and...  am finding that it's not, largely due to the large number of `private` procedures within the `Random` module that essentially need to be replicated in this helper module in order to make use of them.  I'm still looking to see whether I think my proposed approach is a net win (vs. just creating a local copy of the current `Random` module) and will get back to you when I have an answer.

**Comment by bradcray** [comment]
OK, I got a bounded integer fillRandom() working, which you can take a look at on this branch:  https://github.com/bradcray/arkouda/tree/boundedFillRandom

The approach I took:
* updated the Makefile and added compat/ directories as suggested above (where I called my helper module `RandomHelp`)
* added `use RandomHelp` to `RandArray.chpl`
* added some (completely bogus) calls to `fillRandom()` that pass in bogus min/max values
* realized that there isn't a standalone fillRandom() in the Random module that takes min/max values (though there probably should be), so borrowed the logic from the existing standalone fillRandom() calls and made my own version taking bounds in
* copied over the new fillRandom() and iterate() methods from https://github.com/chapel-lang/chapel/pull/19540/files
* then started working through unresolved errors which were due to the large number of private routines used to implement these calls by copying and pasting the routines over to this file.

I think that this is a _reasonable_ approach, though a big question is whether I got all of the helper functions necessary for various code paths to work (e.g., will element types other than integers work, or will I need to copy over additional code?)

Given the amount of unanticipated picking and choosing I had to do to get it to work, it may be that the better approach would be to just copy the version of `Random.chpl` from Chapel's `main` branch into the `compat/lt-127` path where I'd probably preserve the name `Random.chpl` rather than using a different module name so that no code needs to change after 1.27 is released and no `use`s need to change.  The only downside of this approach is that the following warning will be generated (which confirms that the right thing is happening):

```
warning: Ambiguous module source file -- using /path/to/arkouda//src/compat/lt-127/Random.chpl over $CHPL_HOME/modules/standard/Random.chpl
```

Note that in this approach, we'd still need to define a standalone bounded version of fillRandom() (which I think we should do on our main development branch as well).

**Comment by Ethan-DeBandi99** [comment]
@bradcray - Thanks for digging into this. What you are saying makes sense based on what I was seeing. I think I might have been tracking things down incorrectly. I will take a look at this tomorrow. I needed to change gears for today and let you know what I come up with. Again, I really appreciate the help!

**Comment by bradcray** [comment]
This issue just got epsilon easier in that this month's Chapel 1.27.0 should have a free-standing `fillRandom()` routine that takes in bounds thanks to https://github.com/chapel-lang/chapel/pull/19927.  Of course, a compatibility version would still be required for Chapel 1.26.0 to keep two versions of Chapel working.
